api/src/main/java/org/apache/iceberg/transforms/ProjectionUtil.java

```diff
@@ -203,4 +275,24 @@ private ProjectionUtil() {}
         return null;
     }
   }
+
+  static <S, T> UnboundPredicate<T> truncateArrayStrict(
+      String name, BoundPredicate<S> pred, Transform<S, T> transform) {
+    S boundary = pred.literal().value();
+    switch (pred.op()) {
+      case LT:
+      case LT_EQ:
+        return predicate(Expression.Operation.LT, name, transform.apply(boundary));
```

Comment:
Is this correct? Consider `arr < ["a", "b", "c", "d"]`. Truncate length 3 would produce `arr < ["a", "b", "c"]`, which doesn't include the original bound.

This is why for stats truncation, we truncate and then increment the last position, producing `arr < ["a", "b", "d"]`. But I don't think we can do that unless we have a way to increment. I think this has to return `null` instead.

Reply:
Actually, this could use the same methods, like `UnicodeUtil.truncateStringMax`, at least for Strings and ByteBuffers. See https://github.com/apache/incubator-iceberg/blob/master/parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java#L199-L221

Reply:
The way I thought of this, say we have the following partitions with values

`"aba" -> "abab", "abae", "abaf"`
`"abc" -> "abca", "abcd", "abce", "abcf", "abcg"`

Then `< "abcf"` would definitely be non conclusive for the partition `"abc"` and it should be the partitions `< "abc"`. Even `<"abcd"` should exclude the same partition and would only include the lower partitions i.e. `<"abc"`. 

Wouldn't projecting `<"abcd"` to `<"abd"` be wrong since it's including the entire partition of `"abc"` which wouldn't be true for some values like `"abce"`?

Reply:
(Reflected here https://github.com/apache/incubator-iceberg/pull/332/files#diff-f4be406e8a0e2a766149e6c9fe56142fR264)

Reply:
You're right about the `<` case since this is the strict transform.